src/main/scala/analysis/MemoryModelMap.scala

```diff
@@ -1,32 +1,49 @@
 package analysis
 
 import analysis.*
-import ir.BitVecLiteral
+import ir.{BitVecLiteral, Procedure}
 
 import scala.collection.mutable
 
 // Define a case class to represent a range
-case class RangeKey(var start: BigInt, var end: BigInt):
+case class RangeKey(var start: BigInt, var end: BigInt) extends Ordered[RangeKey]:
   def size(): BigInt = end - start
+  override def compare(that: RangeKey): Int = {
+    if (start < that.start) -1
+    else if (start > that.start) 1
+    else 0
+  }
 
 case class RegionToRangesMap():
-  val stackMap: mutable.Map[RangeKey, StackRegion] = mutable.Map()
-  val heapMap: mutable.Map[RangeKey, HeapRegion] = mutable.Map()
-  val dataMap: mutable.Map[RangeKey, DataRegion] = mutable.Map()
+  val stackMap: mutable.Map[RangeKey, StackRegion] = mutable.TreeMap()
+  val sharedStackMap: mutable.Map[Procedure, mutable.TreeMap[RangeKey, StackRegion]] = mutable.Map[Procedure, mutable.TreeMap[RangeKey, StackRegion]]()
+  val heapMap: mutable.Map[RangeKey, HeapRegion] = mutable.TreeMap()
+  val dataMap: mutable.Map[RangeKey, DataRegion] = mutable.TreeMap()
 
 // Custom data structure for storing range-to-object mappings
 class MemoryModelMap {
   //private val functions = scala.collection.mutable.Map[CfgFunctionExitNode, RegionToRangesMap]()
   private val rangeMap = RegionToRangesMap()
   private val MAX_BIGINT: BigInt = BigInt(Long.MaxValue)
   private val contextStack = mutable.Stack.empty[List[StackRegion]]
-  private val allStacks = mutable.Map[String, List[StackRegion]]()
-
-  // Add a range and object to the mapping
-  def add(offset: BigInt, region: MemoryRegion): Unit = {
+  private val sharedContextStack = mutable.Stack.empty[List[StackRegion]]
+  private val localStacks = mutable.Map[String, List[StackRegion]]()
+  private val sharedStacks = mutable.Map[String, List[StackRegion]]().withDefaultValue(List.empty[StackRegion])
```

Comment:
withDefaultValue produces some confusing behaviour and should be avoided

Reply:
It is easy to do weird things with it like mutate the default value instead of mutating the value for a key that hadn't been added, and some operations just confusingly ignore the default.

The alternative is to just check if a key exists in the map before attempting to update its value. This may be a little more verbose but it is worth it.